Complete refactoring of corgi-stateline

theophilusx opened the pull request:
- Change the face to mode-line rather than mode-line-active as that
  face only seems to be in Emacs 29 (development versions)
- Modified the code so that any previous face mapping is removed before
  adding a new face mapping
- Changed the package to define a new minor mode called corig-stateline-mode.
  This makes it easier for users to enable and disable. Also adds a minor mode
  indicator of 'StLi', which users can remove using diminish (or similar).

Probably need to tweak the default color settings. Might be good to start
by modifying the custom definitions for the colors so that it uses different
color sets for dark and light themes.

theophilusx committed:
Complete refactoring of corgi-stateline

- Change the face to mode-line rather than mode-line-active as that
  face only seems to be in Emacs 29 (development versions)
- Modified the code so that any previous face mapping is removed before
  adding a new face mapping
- Changed the package to define a new minor mode called corig-stateline-mode.
  This makes it easier for users to enable and disable. Also adds a minor mode
  indicator of 'StLi', which users can remove using diminish (or similar).

Probably need to tweak the default color settings. Might be good to start
by modifying the custom definitions for the colors so that it uses different
color sets for dark and light themes.

plexus commented:
Thanks a lot @theophilusx, I like the direction you're going. I think turning it into a minor mode makes a lot of sense.

Some remarks

- It seems the face remappings are per-buffer, so we probably want to use `setq-local` for the cookie? (From the docs: "Add a face remapping entry of FACE to SPECS in the current buffer."
- Can we omit the `message` calls when toggling on/off? Seems like more noise than signal.
- Same thing for the lighter. Maybe I'm missing something but I only see visual noise for little to no value. Can we just remove it (set it to blank), instead of making more work for ourselves and others having to turn it off again?
- I think it would make sense to already load and enable this in `corgi-editor`

theophilusx commented:
> It seems the face remappings are per-buffer, so we probably want to use setq-local for the cookie? (From the docs: "Add a face remapping entry of FACE to SPECS in the current buffer."

Will make the change. Was thinking very similar as well.

> Can we omit the message calls when toggling on/off? Seems like more noise than signal.

Oops - debugging code I forgot to remove! Will do so.

> Same thing for the lighter. Maybe I'm missing something but I only see visual noise for little to no value. Can we just remove it (set it to blank), instead of making more work for ourselves and others having to turn it off again?

Yep, can do that. I'm not a fan either. 

>I think it would make sense to already load and enable this in corgi-editor

My only concern with doing that is what impact it will have on users who have a use-package corgi-stateline in their init.el file. It probably won't cause any errors, but could be confusing if they decide they don't want that mode. They will likely remove the corgi-stateline use package form their init.el and wonder why it still gets turned on. Also, if they decide they don't want it, it is harder for them to turn if off permanently (they would have to toggle it off in their init.el). Leaving it as a use-package call in their init.el means they can just comment it out or change it not to turn on initially (but be there should they want it). 

The downside of course is anyone who updates and does not check the changelog will not know it is now a minor mode and it won't be turned on for them as they won't have the new init.el from an updated sample config.

plexus commented:
> users who have a use-package corgi-stateline in their init.el file

That's a good point. Note that for the policy packages like corgi-editor I don't think we can or should commit to always keeping backwards compatibility. People can freeze them with straight or by inlinging them. Of course we also don't want to introduce breakage for the sake of it, and want to provide guidance for people to update their own init.el with each release.

In this case I think you're right for the time being it's probably prudent to keep the status quo and keep it as a separate package. We can still roll it into corgi-editor later on.

theophilusx committed:
Minor tweaks to corgi-stateline-mode

- Make corgi-stateline-remap-cookie buffer local
- Remove minor mode indicator
- Remove messages to make mode less 'chatty'.

theophilusx committed:
Cherry pick of some updates from plexus

- Add global-corgi-stateline-mode
- Add unmap-mode-line-faces
- Remove normal mode color custom variables and remove setting of
  different faces for normal mode (just use theme default).

Note: removed setq-local as I don't beleive it is necesary because
the corgi-stateline-remap-cookie is being defined with defvar-local.
This means that on the first setq, a buffer local copy is created and
all references and setq within that buffer will work on the buffer local
version of the variable.

theophilusx commented:
@plexus I've updated the PR and incorporated some of your changes. In the process, I have seen a couple of unrelated key binding errors and have encountered a few problems with the current ivy/counsel setup. I'm going to spin off a new branch and try to clean these up next as a preliminary step to whatever we decide re: corgi-completions or similar.
